### PR TITLE
Add org-level installation support #617

### LINKED ISSUE #617: Org Apps support
References:
* https://github.com/slackapi/node-slack-sdk/pull/1046
* https://github.com/slackapi/python-slack-sdk/pull/877

### Category (place an `x` in each of the `[ ]`)

* [x] **bolt** (Bolt for Java)
* [ ] **bolt-{sub modules}** (Bolt for Java - optional modules)
* [ ] **slack-api-client** (Slack API Clients)
* [ ] **slack-api-model** (Slack API Data Models)
* [ ] **slack-api-*-kotlin-extension** (Kotlin Extensions for Slack API Clients)
* [ ] **slack-app-backend** (The primitive layer of Bolt for Java)

### Requirements

Please make sure if this topic is specific to this SDK. For general questions/issues about Slack API platform or its server-side, could you submit questions at https://my.slack.com/help/requests/new instead. :bow:

Please read the [Contributing guidelines](https://github.com/slackapi/java-slack-sdk/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to the those rules.

### PR DESCRIPTION
**Pull request by seratch**
This pull request adds a new feature called "Org Apps", which enables Slack apps to be installed into the Enterprise Grid organization, not into each workspace under an organization. This fixes #617 

### Category (place an `x` in each of the `[ ]`)

* [x] **bolt** (Bolt for Java)
* [ ] **bolt-{sub modules}** (Bolt for Java - optional modules)
* [x] **slack-api-client** (Slack API Clients)
* [x] **slack-api-model** (Slack API Data Models)
* [ ] **slack-api-*-kotlin-extension** (Kotlin Extensions for Slack API Clients)
* [x] **slack-app-backend** (The primitive layer of Bolt for Java)

## Requirements

Please read the [Contributing guidelines](https://github.com/slackapi/java-slack-sdk/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to the those rules.

### CODE REVIEW

**Review by stevengill — APPROVED** [review]

**Comment by stevengill** [comment]
LGTM.

Only thing I'm not sure I saw was that during oauth, for org level apps, we redirect to a url where they manage which workspaces to add the newly added app. See https://github.com/slackapi/python-slack-sdk/pull/877/files#diff-a8925230cbd419395f02001ad5fb7363970482b5cce20e6b24a4804d1b82947cR32.

**Comment by seratch** [comment]
I will release v1.3.0 including #622 within a few hours.

**Comment by seratch** [comment]
@stevengill Thanks for reviewing this. As of v1.3, Bolt for Java does not offer the default page rendering features for OAuth install/oauth_redirect URLs but now I'm working on it and will push the changes as part of this PR.

File: bolt/src/main/java/com/slack/api/bolt/context/Context.java [hunk 1]
```diff
@@ -68,11 +73,19 @@
     protected final Map<String, String> additionalValues = new HashMap<>();
 
     public MethodsClient client() {
-        return getSlack().methods(botToken);
+        if (isEnterpriseInstall()) {
+            return getSlack().methods(botToken, teamId);
```
Comment (seratch) [inline bolt/src/main/java/com/slack/api/bolt/context/Context.java:77]: automatically pass teamId to API calls as necessary

File: bolt/src/main/java/com/slack/api/bolt/model/Installer.java [hunk 2]
```diff
@@ -25,6 +25,22 @@
...
 
+    // ---------------------------------
+    // Org-level installation
+    // ---------------------------------
+
+    Boolean getIsEnterpriseInstall();
+
+    void setIsEnterpriseInstall(Boolean isEnterpriseInstall);
+
+    String getEnterpriseUrl();
```
Comment (seratch) [inline bolt/src/main/java/com/slack/api/bolt/model/Installer.java:36]: As @stevengill and I discussed, we decided to rename `orgDashboardGrantAccess` to `enterpriseUrl` for more clarity.

**Review by seratch — COMMENTED** [review]
a few comments for reviewers

File: bolt/src/main/java/com/slack/api/bolt/service/builtin/oauth/view/default_impl/OAuthDefaultRedirectUriPageRenderer.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        String url = completionUrl;
+        if (url == null || url.isEmpty()) {
+            if (installer.getIsEnterpriseInstall() != null
+                    && installer.getIsEnterpriseInstall()
+                    && installer.getEnterpriseUrl() != null
+                    && installer.getAppId() != null
+            ) {
+                // org-level installation
+                String rootUrl = installer.getEnterpriseUrl(); // https://{org domain}.enterprise.slack.com/
+                url = rootUrl + "manage/organization/apps/profile/" + installer.getAppId() + "/workspaces/add";
```
Comment (stevengill) [inline bolt/src/main/java/com/slack/api/bolt/service/builtin/oauth/view/default_impl/OAuthDefaultRedirectUriPageRenderer.java:54]: Looks good!

**Review by stevengill — APPROVED** [review]

**Comment by seratch** [comment]
Thanks for your review, Steve! I will be merging this PR later today. Let me know if you have comments / feedback on this PR by then.

**Comment by seratch** [comment]
>I will be merging this PR later today

I forgot to mention the context here. The `main` branch is already for the next minor version (1.4.0) that won't be released without org-level installation support. So, we can safely merge this PR into main branch and I will continue working on other stuff we will include in v1.4 such as rate-limits support for SCIM API.

In the case where we need to apply changes to 1.3.x, we can use the maintenance branch for the minor version series.

**Pull request merged**